### PR TITLE
OM-1341 Analytics & Enhancements

### PR DESCRIPTION
**Pull request by emessiha**
Main changes are:
- Adding user consent request and analytics
- Some minor bug fixing with clearing Education on content on selecting root node in the file tree
- Refactoring of Package Structure and some class/variable names for clarity
- Tracking info has been verified/communicated with Marketing

### CODE REVIEW

File: src/main/java/com/reshiftsecurity/analytics/AnalyticsActionCategory.java [hunk 1]
```diff
@@ -20,11 +20,19 @@
 
 package com.reshiftsecurity.analytics;
 
-public enum AnalyticsActionType {
-    SCAN, CLICK;
+public enum AnalyticsActionCategory {
+    Code_view,
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/analytics/AnalyticsActionCategory.java]: capital 'V' for consistency

File: src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+    public void recordMetric(AnalyticsAction action, Integer value) {
+        this.entries.add(new AnalyticsEntry(action, value));
+        this.processActions();
+    }
+
+    private static String bytesToHex(byte[] hash) {
+        StringBuffer hexString = new StringBuffer();
+        for (int i = 0; i < hash.length; i++) {
+            String hex = Integer.toHexString(0xff & hash[i]);
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java]: just an improvement on readability these three lines can be
`hexString.append(String.format("%02X", hash[i]));`

File: src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    private String getUserIdentifier() {
+        try {
+            InetAddress ip = InetAddress.getLocalHost();
+            NetworkInterface network = NetworkInterface.getByInetAddress(ip);
+            byte[] mac = network.getHardwareAddress();
+
+            StringBuilder sb = new StringBuilder();
+            for (int i = 0; i < mac.length; i++) {
+                sb.append(String.format("%02X%s", mac[i], (i < mac.length - 1) ? "-" : ""));
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java:122]: if this is going to be a hash to identify the user, why do you need the '-' in between the hex values?

**Review by qcorporation — APPROVED** [review]
looks ok

File: src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java (hunk 3 above)
Comment (emessiha) [reply src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java:122]: From the docs it seemed this is the format they are expecting: https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters#cid

Comment (qcorporation) [reply src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java:122]: sorry I don't understand how this is the same, the mac address, in this context is formatted as xx-xx-xx-xx-xx-xx, but after this, the sb is hashed by sha-256, and then returned. So in this sense adding "-" doesn't really do anything

The documentation that you outlined says the format of the cid
cid=35009a79-1a05-49d7-b876-2b884d0f825b,
so this is a different format from the mac address string value above and also different than the return when you do bytesToHex()

so, I'm not sure how the return of this method gets put into the cid value

Comment (emessiha) [reply src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java:122]: oh good catch! lol yep you're right. The hyphens are not needed anymore lol will remove :)
Seems like they are either not too strict on that format

Comment (qcorporation) [reply src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java:122]: try checking if you can fetch a user and their data, I'd explore google bigquery to see if you can get all elements for that user identifier. If I was google, and a client doesn't conform to those standards, I'd silently fail 😱

Comment (emessiha) [reply src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java:122]: ended up using "uid" which gives more freedom for the format: https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters#uid

**Pull request merged**